builtins/match.go

```diff
@@ -34,9 +38,65 @@ func matchRecord(a, b *runtime.Record) (bool, error) {
 			}
 		} else {
 			return false, runtime.Errorf("Value mismatch: field %s in second Record not found in first", k)
+		}
+	}
+	return true, nil
+}
 
+func matchRecordOf(val, pat []runtime.Object) (bool, error) {
+	i, back_i := 0, -1
+	j, back_j := 0, -1
+	for i < len(val) && j < len(pat) {
+		valueAtI := val[i]
+		patternAtJ := pat[j]
+		i++
+		j++
+
+		switch patternAtJ {
+		case runtime.AnyOrNone:
+			if j == len(pat) { // Optimize trailing * case
+				return true, nil
+			}
+			back_j = j // Pattern Element after *
+			i--        // Allow zero-length match
+			back_i = i // First Value Element which could be matched with that *
+		default: // Literal character or ?
+			if ok, _ := match(valueAtI, patternAtJ); ok {
+				if j == len(pat) && i == len(val) { // if last element in both lists matched
+					return true, nil
+				}
+				break
+			}
+			if back_j < 0 {
```

Comment:
I find the complexity of this algorithm should be reduced. Its complexity is currently 12, while the original algorithm had only 8.

Maybe there's a way reduce or even flatten some conditions?